pandas/core/internals/managers.py

```diff
@@ -1745,10 +1745,12 @@ def as_array(
                     dtype=dtype,
                     na_value=na_value,
                 ).reshape(blk.shape)
+            elif na_value is lib.no_default:
+                arr = np.asarray(blk.get_values(dtype))
             else:
-                arr = np.asarray(blk.get_values())
-                if dtype:
-                    arr = arr.astype(dtype, copy=False)
+                arr = np.empty(self.shape, dtype=dtype)
```

Comment:
todo - check what happens if `dtype=None`, would this default to the wrong dtype?

Reply:
indeed,
```
pd.DataFrame({'a': ['a', 'b', 'c']}).to_numpy(na_value='e')
```
fails